diff --git a/vendor/core/fastboot/fastboot.bash b/vendor/core/fastboot/fastboot.bash
index 911071ae..e88bcf27 100644
--- a/vendor/core/fastboot/fastboot.bash
+++ b/vendor/core/fastboot/fastboot.bash
@@ -15,7 +15,7 @@
 # limitations under the License.
 #
 
-_fastboot() {
+_termux_fastboot() {
     if ! check_type "$1" >/dev/null; then
         return
     fi
@@ -67,7 +67,7 @@ _fastboot() {
             COMPREPLY=( $(compgen -W "$OPTIONS $COMMAND" -- "$cur") )
             ;;
         OPT_SERIAL_ARG)
-            local devices=$(command fastboot devices 2> /dev/null | awk '{ print $1 }')
+            local devices=$(command termux-fastboot devices 2> /dev/null | awk '{ print $1 }')
             COMPREPLY=( $(compgen -W "${devices}" -- ${cur}) )
             ;;
         OPT_SLOT_ARG)
@@ -81,7 +81,7 @@ _fastboot() {
                 i=$((i+1))
                 case "${cur}" in
                     flash)
-                        _fastboot_cmd_flash "$serial" $i
+                        _termux_fastboot_cmd_flash "$serial" $i
                         ;;
                     reboot)
                         if [[ $COMP_CWORD == $i ]]; then
@@ -90,7 +90,7 @@ _fastboot() {
                         fi
                         ;;
                     update)
-                        _fastboot_cmd_update "$serial" $i
+                        _termux_fastboot_cmd_update "$serial" $i
                         ;;
                 esac
             fi
@@ -100,7 +100,7 @@ _fastboot() {
     return 0
 }
 
-_fastboot_cmd_flash() {
+_termux_fastboot_cmd_flash() {
     local serial i cur
     local partitions
 
@@ -112,11 +112,11 @@ _fastboot_cmd_flash() {
         partitions="boot bootloader dtbo init_boot modem odm odm_dlkm oem product pvmfw radio recovery system system_dlkm vbmeta vendor vendor_dlkm vendor_kernel_boot"
         COMPREPLY=( $(compgen -W "$partitions" -- $cur) )
     else
-        _fastboot_util_complete_local_file "${cur}" '!*.img'
+        _termux_fastboot_util_complete_local_file "${cur}" '!*.img'
     fi
 }
 
-_fastboot_cmd_update() {
+_termux_fastboot_cmd_update() {
     local serial i cur
 
     serial=$1
@@ -124,10 +124,10 @@ _fastboot_cmd_update() {
 
     cur="${COMP_WORDS[COMP_CWORD]}"
 
-    _fastboot_util_complete_local_file "${cur}" '!*.zip'
+    _termux_fastboot_util_complete_local_file "${cur}" '!*.zip'
 }
 
-_fastboot_util_complete_local_file() {
+_termux_fastboot_util_complete_local_file() {
     local file xspec i j IFS=$'\n'
     local -a dirs files
 
@@ -176,7 +176,7 @@ _fastboot_util_complete_local_file() {
 }
 
 if [[ $(check_type compopt) == "builtin" ]]; then
-    complete -F _fastboot fastboot
+    complete -F _termux_fastboot termux-fastboot
 else
-    complete -o nospace -F _fastboot fastboot
+    complete -o nospace -F _termux_fastboot termux-fastboot
 fi
diff --git a/vendor/core/fastboot/fastboot.cpp b/vendor/core/fastboot/fastboot.cpp
index 6b9e493e..65528246 100644
--- a/vendor/core/fastboot/fastboot.cpp
+++ b/vendor/core/fastboot/fastboot.cpp
@@ -27,6 +27,7 @@
  */
 
 #include "fastboot.h"
+#include "termux_adb.h"
 
 #include <errno.h>
 #include <fcntl.h>
@@ -2349,6 +2350,8 @@ int FastBootTool::Main(int argc, char* argv[]) {
 
     if (argc == 0 && !fp->wants_wipe && !fp->wants_set_active) syntax_error("no command");
 
+    termuxadb::start();
+
     if (argc > 0 && !strcmp(*argv, "devices")) {
         list_devices();
         return 0;
diff --git a/vendor/core/fastboot/main.cpp b/vendor/core/fastboot/main.cpp
index 35f42189..b39a7dc4 100644
--- a/vendor/core/fastboot/main.cpp
+++ b/vendor/core/fastboot/main.cpp
@@ -27,8 +27,13 @@
  */
 
 #include "fastboot.h"
+#include "termux_adb.h"
 
 int main(int argc, char* argv[]) {
+    if (termuxadb::sendfd()) {
+        return 0;
+    }
+
     FastBootTool fb;
     return fb.Main(argc, argv);
 }
diff --git a/vendor/core/fastboot/termux_adb.h b/vendor/core/fastboot/termux_adb.h
new file mode 100644
index 00000000..6987ca3b
--- /dev/null
+++ b/vendor/core/fastboot/termux_adb.h
@@ -0,0 +1,61 @@
+#pragma once
+
+#include <dirent.h>
+#include <fcntl.h>
+#include <stdarg.h>
+#include <unistd.h>
+
+#include <string>
+
+extern "C" {
+    DIR *termuxadb_opendir(const char *name);
+    int termuxadb_closedir(DIR *dirp);
+    struct dirent *termuxadb_readdir(DIR *dirp);
+
+    int termuxadb_open(const char* path, int options);
+    int termuxadb_create(const char* path, int options, int mode);
+    int termuxadb_close(int fd);
+
+    bool termuxadb_sendfd();
+    void fastboot_start();
+}
+
+namespace termuxadb {
+    static inline DIR *opendir(const char *name) {
+        return termuxadb_opendir(name);
+    }
+
+    static inline int closedir(DIR *dirp) {
+        return termuxadb_closedir(dirp);
+    }
+
+    static inline struct dirent *readdir(DIR *dirp) {
+        return termuxadb_readdir(dirp);
+    }
+
+    static inline int open(std::string_view path, int options, ...) {
+        std::string zero_terminated(path.begin(), path.end());
+        if ((options & O_CREAT) == 0) {
+            return TEMP_FAILURE_RETRY(termuxadb_open(zero_terminated.c_str(), options));
+        } else {
+            int mode;
+            va_list args;
+            va_start(args, options);
+            mode = va_arg(args, int);
+            va_end(args);
+            return TEMP_FAILURE_RETRY(termuxadb_create(zero_terminated.c_str(), options, mode));
+        }
+    }
+
+    static inline int close(int fd) {
+        return termuxadb_close(fd);
+    }
+
+    static inline bool sendfd() {
+        return termuxadb_sendfd();
+    }
+
+    static inline void start() {
+        fastboot_start();
+    }
+}
diff --git a/vendor/core/fastboot/usb_linux.cpp b/vendor/core/fastboot/usb_linux.cpp
index 03af8f76..edd5c980 100644
--- a/vendor/core/fastboot/usb_linux.cpp
+++ b/vendor/core/fastboot/usb_linux.cpp
@@ -36,6 +36,7 @@
 #include <string.h>
 #include <sys/ioctl.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
 #include <sys/types.h>
 #include <unistd.h>
 
@@ -49,6 +50,7 @@
 #include <memory>
 #include <thread>
 
+#include "termux_adb.h"
 #include "usb.h"
 #include "util.h"
 
@@ -152,7 +154,7 @@ static int check(void *_desc, int len, unsigned type, int size)
     return 0;
 }
 
-static int filter_usb_device(char* sysfs_name,
+static int filter_usb_device(int fd,
                              char *ptr, int len, int writable,
                              ifc_match_func callback,
                              int *ept_in_id, int *ept_out_id, int *ifc_id)
@@ -186,6 +188,26 @@ static int filter_usb_device(char* sysfs_name,
     info.dev_protocol = dev->bDeviceProtocol;
     info.writable = writable;
 
+    struct stat st;
+    char pathbuf[128];
+    char link[256];
+    char *sysfs_name = nullptr;
+
+    if (!fstat(fd, &st) && S_ISCHR(st.st_mode)) {
+        snprintf(pathbuf, sizeof(pathbuf), "/sys/dev/char/%d:%d",
+                major(st.st_rdev), minor(st.st_rdev));
+        ssize_t link_len = readlink(pathbuf, link, sizeof(link) - 1);
+        if (link_len > 0) {
+            link[link_len] = '\0';
+            const char* slash = strrchr(link, '/');
+            if (slash) {
+                snprintf(pathbuf, sizeof(pathbuf),
+                        "%s", slash + 1);
+                sysfs_name = pathbuf;
+            }
+        }
+    }
+
     snprintf(info.device_path, sizeof(info.device_path), "usb:%s", sysfs_name);
 
     /* Read device serial number (if there is one).
@@ -203,11 +225,11 @@ static int filter_usb_device(char* sysfs_name,
                  "/sys/bus/usb/devices/%s/serial", sysfs_name);
         path[sizeof(path) - 1] = '\0';
 
-        fd = open(path, O_RDONLY);
+        fd = termuxadb::open(path, O_RDONLY);
         if (fd >= 0) {
             int chars_read = read(fd, info.serial_number,
                                   sizeof(info.serial_number) - 1);
-            close(fd);
+            termuxadb::close(fd);
 
             if (chars_read <= 0)
                 info.serial_number[0] = '\0';
@@ -337,30 +359,16 @@ static int read_sysfs_number(const char *sysfs_name, const char *sysfs_node)
     return value;
 }
 
-/* Given the name of a USB device in sysfs, get the name for the same
- * device in devfs. Returns 0 for success, -1 for failure.
- */
-static int convert_to_devfs_name(const char* sysfs_name,
-                                 char* devname, int devname_size)
-{
-    int busnum, devnum;
-
-    busnum = read_sysfs_number(sysfs_name, "busnum");
-    if (busnum < 0)
-        return -1;
-
-    devnum = read_sysfs_number(sysfs_name, "devnum");
-    if (devnum < 0)
-        return -1;
-
-    snprintf(devname, devname_size, "/dev/bus/usb/%03d/%03d", busnum, devnum);
-    return 0;
+static inline bool contains_non_digit(const char* name) {
+    while (*name) {
+        if (!isdigit(*name++)) return true;
+    }
+    return false;
 }
 
-static std::unique_ptr<usb_handle> find_usb_device(const char* base, ifc_match_func callback)
+static std::unique_ptr<usb_handle> find_usb_device(const std::string& base, ifc_match_func callback)
 {
     std::unique_ptr<usb_handle> usb;
-    char devname[64];
     char desc[1024];
     int n, in, out, ifc;
 
@@ -368,42 +376,49 @@ static std::unique_ptr<usb_handle> find_usb_device(const char* base, ifc_match_f
     int fd;
     int writable;
 
-    std::unique_ptr<DIR, decltype(&closedir)> busdir(opendir(base), closedir);
-    if (busdir == 0) return 0;
+    std::unique_ptr<DIR, int(*)(DIR*)> bus_dir(termuxadb::opendir(base.c_str()), termuxadb::closedir);
+    if (!bus_dir) return nullptr;
+
+    while ((de = termuxadb::readdir(bus_dir.get())) != nullptr) {
+        if (contains_non_digit(de->d_name)) continue;
+
+        std::string bus_name = base + "/" + de->d_name;
+
+        std::unique_ptr<DIR, int(*)(DIR*)> dev_dir(termuxadb::opendir(bus_name.c_str()), termuxadb::closedir);
+        if (!dev_dir) continue;
 
-    while ((de = readdir(busdir.get())) && (usb == nullptr)) {
-        if (badname(de->d_name)) continue;
+        while ((de = termuxadb::readdir(dev_dir.get()))) {
+            if (contains_non_digit(de->d_name)) continue;
 
-        if (!convert_to_devfs_name(de->d_name, devname, sizeof(devname))) {
+            std::string dev_name = bus_name + "/" + de->d_name;
 
-//            DBG("[ scanning %s ]\n", devname);
             writable = 1;
-            if ((fd = open(devname, O_RDWR)) < 0) {
+            if ((fd = termuxadb::open(dev_name, O_RDWR)) < 0) {
                 // Check if we have read-only access, so we can give a helpful
                 // diagnostic like "adb devices" does.
                 writable = 0;
-                if ((fd = open(devname, O_RDONLY)) < 0) {
+                if ((fd = termuxadb::open(dev_name, O_RDONLY)) < 0) {
                     continue;
                 }
             }
 
             n = read(fd, desc, sizeof(desc));
 
-            if (filter_usb_device(de->d_name, desc, n, writable, callback, &in, &out, &ifc) == 0) {
+            if (filter_usb_device(fd, desc, n, writable, callback, &in, &out, &ifc) == 0) {
                 usb.reset(new usb_handle());
-                strcpy(usb->fname, devname);
+                strcpy(usb->fname, dev_name.c_str());
                 usb->ep_in = in;
                 usb->ep_out = out;
                 usb->desc = fd;
 
                 n = ioctl(fd, USBDEVFS_CLAIMINTERFACE, &ifc);
                 if (n != 0) {
-                    close(fd);
+                    termuxadb::close(fd);
                     usb.reset();
                     continue;
                 }
             } else {
-                close(fd);
+                termuxadb::close(fd);
             }
         }
     }
@@ -557,7 +572,7 @@ int LinuxUsbTransport::Close()
     fd = handle_->desc;
     handle_->desc = -1;
     if(fd >= 0) {
-        close(fd);
+        termuxadb::close(fd);
         DBG("[ usb closed %d ]\n", fd);
     }
 
@@ -576,7 +591,7 @@ int LinuxUsbTransport::Reset() {
 
 std::unique_ptr<UsbTransport> usb_open(ifc_match_func callback, uint32_t timeout_ms) {
     std::unique_ptr<UsbTransport> result;
-    std::unique_ptr<usb_handle> handle = find_usb_device("/sys/bus/usb/devices", callback);
+    std::unique_ptr<usb_handle> handle = find_usb_device("/dev/bus/usb", callback);
 
     if (handle) {
         result = std::make_unique<LinuxUsbTransport>(std::move(handle), timeout_ms);
